File: typescript/scripts/cross-lang-pipeline-dot.ts
#!/usr/bin/env -S npx tsx
/**
 * Cross-language DOT byte-parity fixture: builds the canonical
 * producer / bounded-buffer / consumer pipeline composed via
 * {@link PetriNet.builder().compose}, and writes the resulting DOT to the
 * path supplied via `--out=<path>` (or `target/cross-lang-pipeline-dot/ts.dot`
 * by default, relative to `typescript/`).
 *
 * Mirrors `java/src/test/java/org/libpetri/export/CrossLangPipelineDot.java`
 * and `rust/libpetri/tests/cross_lang_pipeline_dot.rs`. Consumed by
 * `scripts/cross-lang-dot-parity.sh`.
 *
 * Usage (from repo root):
 *   cd typescript && npx tsx scripts/cross-lang-pipeline-dot.ts --out=/tmp/ts.dot
 */

import { mkdirSync, writeFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { place } from '../src/core/place.js';
import { PetriNet } from '../src/core/petri-net.js';
import { dotExport } from '../src/export/dot-exporter.js';
import { boundedBuffer, consumer, producer } from '../tests/fixtures/subnet-fixtures.js';

const DEFAULT_OUT = 'target/cross-lang-pipeline-dot/ts.dot';

function parseOut(argv: readonly string[]): string {
  for (const arg of argv) {
    if (arg.startsWith('--out=')) return arg.slice('--out='.length);
  }
  return DEFAULT_OUT;
}

function main(): void {
  const producerDef = producer();
  const bufferDef = boundedBuffer(2);
  const consumerDef = consumer();

  const prodInst = producerDef.instantiate('prod');
  const bufInst = bufferDef.instantiate('buf');
  const consInst = consumerDef.instantiate('cons');

  const producerToBuffer = place<string>('producerToBuffer');
  const bufferToConsumer = place<string>('bufferToConsumer');

  const net = PetriNet.builder('pipeline')
    .compose(prodInst, (b) => b.bindPort<string>('output', producerToBuffer))
    .compose(bufInst, (b) =>
      b.bindPort<string>('put', producerToBuffer).bindPort<string>('get', bufferToConsumer),
    )
    .compose(consInst, (b) => b.bindPort<string>('input', bufferToConsumer))
    .build();

  const dot = dotExport(net);

  const __dirname = dirname(fileURLToPath(import.meta.url));
  const tsRoot = resolve(__dirname, '..');
  const outPath = parseOut(process.argv.slice(2));
  const target = resolve(tsRoot, outPath);
  mkdirSync(dirname(target), { recursive: true });
  writeFileSync(target, dot);
}

main();

File: java/src/test/java/org/libpetri/export/CrossLangPipelineDot.java
package org.libpetri.export;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import org.junit.jupiter.api.Test;
import org.libpetri.core.PetriNet;
import org.libpetri.core.Place;
import org.libpetri.fixtures.SubnetFixtures;

/**
 * Cross-language DOT byte-parity fixture: builds the canonical
 * producer / bounded-buffer / consumer pipeline composed via
 * {@link PetriNet.Builder#compose} and emits the resulting DOT to the path
 * named by the {@code libpetri.crossLang.outFile} system property (or to
 * {@code target/cross-lang-pipeline-dot/java.dot} by default).
 *
 * <p>The output is consumed by {@code scripts/cross-lang-dot-parity.sh},
 * which builds the same pipeline in TypeScript and Rust and asserts the
 * three DOT outputs are byte-identical.
 *
 * <p>Implemented as a JUnit test (rather than a {@code main} class) so
 * regular {@code ./mvnw test} runs verify the pipeline at least executes
 * end-to-end. The parity script invokes this single test with
 * {@code -Dtest=CrossLangPipelineDot}.
 */
public class CrossLangPipelineDot {

    /** Default output path, relative to the {@code java/} module root. */
    public static final String DEFAULT_OUT = "target/cross-lang-pipeline-dot/java.dot";

    @Test
    void emitsCanonicalPipelineDot() throws IOException {
        var producerDef = SubnetFixtures.producer();
        var bufferDef = SubnetFixtures.boundedBuffer(2);
        var consumerDef = SubnetFixtures.consumer();

        var prod = producerDef.instantiate("prod");
        var buf = bufferDef.instantiate("buf");
        var cons = consumerDef.instantiate("cons");

        Place<String> producerToBuffer = Place.of("producerToBuffer", String.class);
        Place<String> bufferToConsumer = Place.of("bufferToConsumer", String.class);

        var net = PetriNet.builder("pipeline")
            .compose(prod, b -> b.bindPort("output", producerToBuffer))
            .compose(buf, b -> b.bindPort("put", producerToBuffer)
                                .bindPort("get", bufferToConsumer))
            .compose(cons, b -> b.bindPort("input", bufferToConsumer))
            .build();

        String dot = DotExporter.export(net);

        String outPath = System.getProperty("libpetri.crossLang.outFile", DEFAULT_OUT);
        Path target = Path.of(outPath);
        if (target.getParent() != null) {
            Files.createDirectories(target.getParent());
        }
        Files.writeString(target, dot);
        // Echo to stdout so the parity script can also pipe-capture if it
        // prefers (when -Dlibpetri.crossLang.echo=true is set).
        if (Boolean.getBoolean("libpetri.crossLang.echo")) {
            System.out.print(dot);
        }
    }
}
